File: test/org/cerion/symcalc/expression/FunctionExprTest.java
package org.cerion.symcalc.expression;

import org.cerion.symcalc.expression.function.FunctionExpr;
import org.cerion.symcalc.expression.function.arithmetic.Plus;
import org.cerion.symcalc.expression.function.arithmetic.Subtract;
import org.cerion.symcalc.expression.number.IntegerNum;
import org.cerion.symcalc.expression.number.RealNum;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

public class FunctionExprTest {
    @Test
    public void equals() {
        FunctionExpr f1 = new Plus(new IntegerNum(3), RealNum.Companion.create(5.0));
        FunctionExpr f2 = new Subtract(new IntegerNum(3), RealNum.Companion.create(5.0));
        FunctionExpr f3 = new Plus(new IntegerNum(4), RealNum.Companion.create(5.0));
        FunctionExpr f4 = new Plus(new IntegerNum(3), RealNum.Companion.create(5.0)); // same as f1

        assertNotEquals(f1, f2);
        assertNotEquals(f1, f3);
        assertNotEquals(f2, f3);
        assertEquals(f1, f4);
    }

}

File: test/expression/FunctionExprTest.kt
package org.cerion.symcalc.expression

import org.cerion.symcalc.expression.function.arithmetic.Plus
import org.cerion.symcalc.expression.function.arithmetic.Subtract
import org.cerion.symcalc.expression.function.arithmetic.Times
import org.cerion.symcalc.expression.number.IntegerNum
import org.cerion.symcalc.expression.number.RealNum
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Test

class FunctionExprTest {
    @Test
    fun equals() {
        val f1 = Plus(IntegerNum(3), RealNum.create(5.0))
        val f2 = Subtract(IntegerNum(3), RealNum.create(5.0))
        val f3 = Plus(IntegerNum(4), RealNum.create(5.0))
        val f4 = Plus(IntegerNum(3), RealNum.create(5.0)) // same as f1

        assertNotEquals(f1, f2)
        assertNotEquals(f1, f3)
        assertNotEquals(f2, f3)
        assertEquals(f1, f4)
    }

    @Test
    fun equals_Orderless() {
        assertEquals(Times(IntegerNum(1), IntegerNum(2), IntegerNum(3)), Times(IntegerNum(2), IntegerNum(3), IntegerNum(1)))
    }
}